weed/filer/reader_at_test.go

```diff
@@ -163,3 +163,86 @@ func TestReaderAt1(t *testing.T) {
 	testReadAt(t, readerAt, 1, 10, 10, nil)
 
 }
+
+func testReadAtWithMonitorReadAt(t *testing.T, readerAt *ChunkReadAt, offset int64, size int, expected int, expectedErr error) {
+	data := make([]byte, size)
+	n, err := readerAt.ReadAt(data, offset)
+
+	for _, d := range data {
+		fmt.Printf("%x", d)
+	}
+	fmt.Println()
+
+	if expected != n {
+		t.Errorf("unexpected read size: %d, expect: %d", n, expected)
+	}
+	if err != expectedErr {
+		t.Errorf("unexpected read error: %v, expect: %v", err, expectedErr)
+	}
+
+}
+
+func TestReaderAtRandomWithMonitorReadAt(t *testing.T) {
+
+	visibles := []VisibleInterval{
+		{
+			start:     0,
+			stop:      2,
+			fileId:    "1",
+			chunkSize: 2,
+		},
+		{
+			start:     2,
+			stop:      4,
+			fileId:    "3",
+			chunkSize: 2,
+		},
+		{
+			start:     4,
+			stop:      6,
+			fileId:    "5",
+			chunkSize: 2,
+		},
+		{
+			start:     6,
+			stop:      8,
+			fileId:    "7",
+			chunkSize: 2,
+		},
+		{
+			start:     8,
+			stop:      10,
+			fileId:    "9",
+			chunkSize: 2,
+		},
+	}
+
+	readerAt := &ChunkReadAt{
+		chunkViews:    ViewFromVisibleIntervals(visibles, 0, math.MaxInt64),
+		readerLock:    sync.Mutex{},
+		fileSize:      10,
+		readerCache:   newReaderCache(3, &mockChunkCache{}, nil),
+		readerPattern: NewReaderPattern(),
+	}
+
+	testReadAt(t, readerAt, 0, 10, 10, io.EOF)
+	testReadAt(t, readerAt, 1, 45, 9, io.EOF)
+	testReadAt(t, readerAt, 42, 8, 0, io.EOF)
+	testReadAt(t, readerAt, 3, 6, 6, nil)
+
+	// we don't implement rand read api in test, if c.readerPattern.IsRandomMode() == true, test should get this error
+	readerAt.readerPattern = NewReaderPattern()
+	testReadAtWithMonitorReadAt(t, readerAt, 2, 8, 0, ErrNilFuncNotPermitted)
```

Comment:
the test only checks errors are valid. How the data are checked?

Reply:
Well, This pr aims at the judgement of the RandomRead in SeaweedFS.Mount client.

I read the file read flow of SeaweedFS.Mount, The single read range [start, end] flow is:
```
1. if the read op is a random read, just fetch VolumeServer over HTTP request with Header 'Range: bytes=start-end', return
2. if the read op is not a random read, caching the hole chunk from VolumeServer, return [start, end] content.
```

#### To check the data, I should write a unit test to verify the range interface and file cache interface of the volume server, I think seaweedfs/weed/filer is not the suitable place to add the unit test.

So I write the unit test to cover the case of "Corrective Judgment", and this unit test is not aims at detect the error of ErrNilFuncNotPermitted.
The error ErrNilFuncNotPermitted will never happen beside unit test. 

This error is only use in unit test to detect the flow of code branch execution
```
1. func testReadAtWithMonitorReadAt() will return ErrNilFuncNotPermitted if a doReadAt() is a random read.
2. func testReadAtWithMonitorReadAt() will return nil/io.EOF if a doReadAt() is a none random read.
```
Please let me know if my understanding is wrong :)